tests/core/selectionoptimizer.js

```diff
@@ -0,0 +1,68 @@
+( function() {
+	'use strict';
+
+	bender.editors = {
+		classic: {},
+		divarea: {
+			config: {
+				extraPlugins: 'divarea'
+			}
+		},
+		inline: {
+			creator: 'inline'
+		}
+	};
+
+	// (#3175)
+	var tests = {
+		'test selection optimization case 1': testSelection( {
+			initial: '<p>foo</p><p>[bar</p><p>]baz</p>',
+			expected: '<p>foo</p><p>[bar]</p><p>baz</p>'
+		} ),
+		'test selection optimization case 2': testSelection( {
+			initial: '<p>foo[</p><p>bar]</p><p>baz</p>',
+			expected: '<p>foo</p><p>[bar]</p><p>baz</p>'
+		} ),
+		'test selection optimization case 3': testSelection( {
+			initial: '<p>foo[</p><p>bar</p><p>]baz</p>',
+			expected: '<p>foo</p><p>[bar]</p><p>baz</p>'
+		} ),
+		'test selection optimization case 4': testSelection( {
+			initial: '<p>foo</p><p>[bar</p><p>b]az</p>',
+			expected: '<p>foo</p><p>[bar</p><p>b]az</p>'
+		} ),
+		'test selection optimization case 5': testSelection( {
+			initial: '<p>fo[o</p><p>bar]</p><p>baz</p>',
+			expected: '<p>fo[o</p><p>bar]</p><p>baz</p>'
+		} ),
+		'test selection optimization case 6': testSelection( {
+			initial: '<p>foo</p><p>[bar</p><p>]baz</p>',
+			expected: '<p>foo</p><p>[bar]</p><p>baz</p>'
+		} ),
+		'test selection optimization case 7': testSelection( {
```

Comment:
Some more complex use cases with more block elements (e.g. tables, lists, blockquote) and some nested inline elements (styling spans, bold, etc.) should be also added.